cmd/geth/dbcmd.go

```diff
@@ -684,3 +701,84 @@ func exportChaindata(ctx *cli.Context) error {
 	db := utils.MakeChainDatabase(ctx, stack, true)
 	return utils.ExportChaindata(ctx.Args().Get(1), kind, exporter(db), stop)
 }
+
+func freezerMigrate(ctx *cli.Context) error {
+	stack, _ := makeConfigNode(ctx)
+	defer stack.Close()
+
+	db := utils.MakeChainDatabase(ctx, stack, false)
+	defer db.Close()
+
+	// Check first block for legacy receipt format
+	numAncients, err := db.Ancients()
+	if err != nil {
+		return err
+	}
+	if numAncients < 1 {
+		log.Info("No blocks in freezer to migrate")
+		return nil
+	}
+
+	// Find first block with non-empty receipt
+	firstIdx := uint64(0)
+	emptyRLPList := []byte{192}
+	for i := uint64(0); i < numAncients; i++ {
+		r, err := db.Ancient("receipts", i)
+		if err != nil {
+			return err
+		}
+		if len(r) == 0 {
+			continue
+		}
+		if !bytes.Equal(r, emptyRLPList) {
+			firstIdx = i
+			break
+		}
+	}
+	// Is first non-empty receipt legacy?
+	first, err := db.Ancient("receipts", firstIdx)
+	if err != nil {
+		return err
+	}
+	isFirstLegacy, err := types.IsLegacyStoredReceipts(first)
+	if err != nil {
+		return err
+	}
+	if !isFirstLegacy {
+		log.Info("No legacy receipts to migrate", "number", firstIdx)
+		return nil
+	}
+	log.Info("First legacy receipt", "number", firstIdx)
```

Comment:
Afaict, this is the last time you make any use of `firstIdx`: the rest of the code starts over from `0` and feeds it into the `transformer`. And the transformer will (?) return `stop=true` for the first empty receipt list, so I don't see how this can work?

Reply:
I had to double-check to see why its working :) so an empty receipt list is both legacy and non-legacy and transformer will ask is this legacy? it is so `stop=false`. `stop` will be true only for the first non-empty non-legacy receipt list